### PR TITLE
myst as source format for demos

### PR DESCRIPTION
**Pull request by jhale**
This is work in progress. Currently I am trying to find a way to get the current demos to myst without too much pain.

The key file to look at (the format we would be working in) in the current commit is: python/demo/poisson/experiments/demo_poisson.py.manual.md

So far I am:

- Automatically converting to rst with pylit
- Automatically converting to myst automatically using rst-to-myst
- Manually adding {code-cell} to code blocks ``` ```. This would be the new 'master format'.
- Converting to ipynb, py etc. using jupytext as needed.

Sphinx should support myst with an extension.

### CODE REVIEW

**Review by garth-wells — COMMENTED** [review]
I think this looks very nice. I ran it and made small changes and ran again. Nice to be using simple Markdown.

Is there anyway to auto-generate links from a function in the code to the doc entry for the function?

File: python/demo/poisson/demo_poisson.py [hunk 1]
```diff
@@ -1,204 +1,236 @@
...
+# `False` for the points outside. In our case, we want to say that the
+# points $(x, y)$ such that $x = 0$ or $x = 1$ are inside on the inside
+# of $\Gamma_D$. (Note that because of rounding-off errors, it is often
+# wise to instead specify $x < \epsilon$ or $x > 1 - \epsilon$ where
+# $\epsilon$ is a small number (such as machine precision).)
 #
 # Now, the Dirichlet boundary condition can be created using the class
-# :py:class:`DirichletBC <dolfinx.fem.bcs.DirichletBC>`. A
-# :py:class:`DirichletBC <dolfinx.fem.bcs.DirichletBC>` takes three
+# {py:class}`DirichletBC <dolfinx.fem.DirichletBC>`. A
```
Comment (jorgensd) [inline python/demo/poisson/demo_poisson.py:135]: This documentation is outdated, as we now use the function `dolfinx.fem.dirichletbc` to create a DirichletBC of class
`dolfinx.fem.bcs.DirichletBCMetaClass`

File: python/demo/poisson/demo_poisson.py [hunk 2]
```diff
@@ -1,204 +1,236 @@
...
-
-# The second argument to :py:class:`FunctionSpace
-# <dolfinx.fem.FunctionSpace>` is the finite element family, while the
-# third argument specifies the polynomial degree. Thus, in this case,
-# our space ``V`` consists of first-order, continuous Lagrange finite
-# element functions (or in order words, continuous piecewise linear
-# polynomials).
+# function space $V$ relative to this mesh. We create a rectangular mesh
+# using built-in function provided by class
+# {py:class}`create_rectangle<dolfinx.mesh.create_rectangle>`. In order
```
Comment (jorgensd) [inline python/demo/poisson/demo_poisson.py:100]: This should be a `{py:func}`?

File: python/demo/poisson/demo_poisson.py [hunk 3]
```diff
@@ -1,204 +1,236 @@
...
 # boundary which the condition apply to, and the function space. This
 # boundary part is identified with degrees of freedom in the function
 # space to which we apply the boundary conditions.
 #
 # To identify the degrees of freedom, we first find the facets (entities
 # of dimension 1) that likes on the boundary of the mesh, and satisfies
-# our criteria for `\Gamma_D`. Then, we use the function
-# ``locate_dofs_topological`` to identify all degrees of freedom that is
+# our criteria for $\Gamma_D$. Then, we use the function
+# `locate_dofs_topological` to identify all degrees of freedom that is
```
Comment (jorgensd) [inline python/demo/poisson/demo_poisson.py:145]: Should this be ?
```
{py:func}`locate_dofs_topological <dolfinx.fem.locate_dofs_topological>`
```

File: python/demo/poisson/demo_poisson.py (hunk 2 above)
Comment (jhale) [reply python/demo/poisson/demo_poisson.py:100]: Yes there are lots of mistakes :) This conversion was done (almost) automatically. The purpose of this PR is to get the infrastructure in place, we can look at content on another PR.

**Pull request merged**